src/clientagent/AstronClient.cpp

```diff
@@ -61,13 +70,33 @@ class AstronClient : public Client, public NetworkClient
         Client(config, client_agent), NetworkClient(stream), m_config(config),
         m_clean_disconnect(false), m_relocate_owned(relocate_owned.get_rval(config)),
         m_send_hash(send_hash_to_client.get_rval(config)),
-        m_send_version(send_version_to_client.get_rval(config))
+        m_send_version(send_version_to_client.get_rval(config)),
+        m_heartbeat_timeout(heartbeat_timeout_config.get_rval(config))
     {
         initialize();
     }
 
+    ~AstronClient()
+    {
+      if (m_heartbeat_timer != nullptr) {
+        delete m_heartbeat_timer;
+      }
+    }
+
+    void heartbeat_timeout()
+    {  
+    lock_guard<recursive_mutex> lock(m_client_lock);
+    send_disconnect(CLIENT_DISCONNECT_NO_HEARTBEAT,
+                    "Server timed out while waiting for heartbeat."); 
+    }
+
     void initialize()
     {
+        //If heartbeat, start the heartbeat timer now.
+        if (m_heartbeat_timeout != 0) {
+            m_heartbeat_timer = new Timeout(m_heartbeat_timeout, std::bind(&AstronClient::heartbeat_timeout,this));
```

Comment:
I'm getting *really* picky here, but if you're making edits, put a space before "this"